charmhelpers/fetch/ubuntu.py

```diff
@@ -457,7 +545,55 @@ def _add_apt_repository(spec):
     if '{series}' in spec:
         series = lsb_release()['DISTRIB_CODENAME']
         spec = spec.replace('{series}', series)
-    _run_with_retries(['add-apt-repository', '--yes', spec])
+    # software-properties package for bionic properly reacts to proxy settings
+    # passed as environment variables (See lp:1433761). This is not the case
+    # LTS and non-LTS releases below bionic.
+    _run_with_retries(['add-apt-repository', '--yes', spec],
+                      cmd_env=_env_proxy_settings(['https']))
+
+
+def _env_proxy_settings(selected_settings=None):
+    """
+    Get charm proxy settings from environment variables that correspond to
+    juju-http-proxy, juju-https-proxy and juju-no-proxy (available as of 2.4.2,
+    see lp:1782236) in a format suitable for passing to an application that
+    reacts to proxy settings passed as environment variables. Some applications
+    support lowercase or uppercase notation (e.g. curl), some support only
+    lowercase (e.g. wget), there are also subjectively rare cases of only
+    uppercase notation support. no_proxy CIDR and wildcard support also varies
+    between runtimes and applications as there is no enforced standard.
+
+    Some applications may connect to multiple destinations and expose config
+    options that would affect only proxy settings for a specific destination
+    these should be handled in charms in an application-specific manner.
+
+    :param selected_settings: format only a subset of possible settings
+    """
+    SUPPORTED_SETTINGS = {
+        'http': 'HTTP_PROXY',
+        'https': 'HTTPS_PROXY',
+        'no_proxy': 'NO_PROXY',
+        'ftp': 'FTP_PROXY'
+    }
+    if not selected_settings:
```

Comment:
Please explicitly check for `None` if using it as a sentinel for "parameter not set by caller". e.g.

```python
if selected_settings is None:
    selected_settings = SUPPORTED_SETTINGS
```

This makes it clear that this piece of code is actually doing the default, rather than if the user passed `{}` for example.